### PR TITLE
qtile: Fix QTile restart environment

### PR DESCRIPTION
**Pull request by Rutherther**
### Issue

I stumbled across a bug in the QTile module where if qtile is restarted via their restart mechanism (https://github.com/qtile/qtile/blob/master/libqtile/core/lifecycle.py#L35), and there are extra packages configured in the module, the Python modules won't be in the new environment, and thus lead to `ModuleNotFoundError` if the modules are used in `config.py`.

### Notes

The cause of this is that the executable that is being used for restart is the unwrapped package itself, not the python environment that contains the extra packages.

I would welcome a discussion about possible solutions. My solution is sort of an ad-hoc solution, that works, but doesn't seem ideal to me. It seems to me these are basically the options we have:
- Change restart behavior inside of QTile to call the correct executable
- Make sure the argv[0] for QTile is the python environment wrapper file instead of the QTile unwrapped package 
- pass packages to the unwrapped package to make module available inside of QTile unwrapped package (what I did)

The other issue is good implementation of one of these options.

I am open to suggestions and willing to implement them myself.

### Current solution

Currently I implemented a solution where the package going to the module is overriden, and `extraPackages` are passed into the `propagatedBuildInputs` of the unwrapped package. This seems to be working fine. I have also removed the wrapping python environment since it seems to have become obsolete by the change I introduced.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by arjan-s** [comment]
This PR is obsolete since https://github.com/NixOS/nixpkgs/pull/307011 was merged with mostly the same code.

**Comment by Rutherther** [comment]
I disagree it's obsolete. This PR solves different issue. Yes, including with it also finalPackage that is now merged as well, so part of this PR should be adapted. The underlying issue is still present even with #307011, since it still uses wrapper environment instead of overriding the inputs of qtile itself.

**Comment by Rutherther** [comment]
Adapted the PR to new #307011 changes.

**Comment by arjan-s** [comment]
Interesting, thanks. And apologies for my too definitive word choice. I couldn't test this myself because I'm not running on xorg but on wayland (where restarting is not possible).

Anyway, I'm looking to change this code anyway in order to support actual session desktop files, so please hold on a bit longer. I hope the changes I want to make will help this case too.

**Comment by Rutherther** [comment]
TL;DR of the problem is that qtile `sys.executable` is set to one coming from `pkgs.qtile` instead of the `finalPackage` (final environment) that has the packages such as qtile-extras.

**Comment by Sigmanificient** [comment]
Qtile `xorg` user here.

Current my config is running on nixpkgs unstable at [this commit](https://github.com/NixOS/nixpkgs/5710852ba686cc1fd0d3b8e22b3117d43ba374c2) (2024-05-21).
Using `pkill -f qtile -SIGUSR2` to restart, seems to work fine
```
...
2024-05-31 13:55:42,431 WARNING libqtile lifecycle.py:_atexit():L33 Restarting Qtile with os.execv(...)
```

However, I updated my flake inputs to [this commit](https://github.com/NixOS/nixpkgs/ef54ee70744c03cfa5ed7cea4e3d96d84cdcb39a) (2024-05-31) and the same manipulation yield the following error:
```
...
2024-05-31 13:55:42,431 WARNING libqtile lifecycle.py:_atexit():L33 Restarting Qtile with os.execv(...)
2024-05-31 14:07:14,632 WARNING libqtile lifecycle.py:_atexit():L33 Restarting Qtile with os.execv(...)
2024-05-31 14:07:16,705 WARNING libqtile lifecycle.py:_atexit():L33 Restarting Qtile with os.execv(...)
``` 
(accidentally restart it twice)

---

With 
```diff
 windowManager.qtile = {
   enable = true;
   backend = "x11";
+ extraPackages = ps: [ ps.qtile-extras ];
 };
```
it seems to work fine too
```
...
2024-05-31 14:07:16,705 WARNING libqtile lifecycle.py:_atexit():L33 Restarting Qtile with os.execv(...)
2024-05-31 14:12:13,115 WARNING libqtile lifecycle.py:_atexit():L33 Restarting Qtile with os.execv(...)
```

Am i missing anything?

**Comment by Sigmanificient** [comment]
Oh you right, added `AnalogClock` from `qtile_extras.widgets`
```
2024-05-31 14:19:54,244 ERROR libqtile manager.py:restart():L250 Preventing restart because of a configuration error:
Traceback (most recent call last):
  File "/nix/store/jn8bd651xg70rpv9d8b5w5a17ayy4058-python3.11-qtile-0.25.0/lib/python3.11/site-packages/libqtile/core/manager.py", line 248, in restart
    self.config.load()
  File "/nix/store/jn8bd651xg70rpv9d8b5w5a17ayy4058-python3.11-qtile-0.25.0/lib/python3.11/site-packages/libqtile/confreader.py", line 129, in load
    self._reload_config_submodules(path)
  File "/nix/store/jn8bd651xg70rpv9d8b5w5a17ayy4058-python3.11-qtile-0.25.0/lib/python3.11/site-packages/libqtile/confreader.py", line 118, in _reload_config_submodules
    importlib.reload(module)
  File "/nix/store/7hnr99nxrd2aw6lghybqdmkckq60j6l9-python3-3.11.9/lib/python3.11/importlib/__init__.py", line 169, in reload
    _bootstrap._exec(spec, module)
  File "<frozen importlib._bootstrap>", line 621, in _exec
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/home/sigmanificient/.config/qtile/core/bar.py", line 1, in <module>
    from qtile_extras import widget
ModuleNotFoundError: No module named 'qtile_extras'
```

**Comment by Sigmanificient** [comment]
It's a bit hard to test the fix as is for me (without modifying some stuff on my config), i'll write a nixos wm test for it

**Comment by Sigmanificient** [comment]
I made a series of patches on the qtile nixos vm test that showcase better the issue:

- move `qtile.nix` -> `qtile/default.nix`
- reflect `qtile.nix` -> `qtile/default.nix` in all-tests.nix

**nixos/tests/qtile.nix:**

- add the default config with a small patch featuring `qtile_extras`
- add a test running `pkiill -f qtile -SIGUSR2` to force qtile restart
nix-build -A qtile-unwrapped.passthru.tests.qtile

**qtile/add-widget.patch:**

- a small patch file to add the analogue clock from `qtile-extras` in the bar:

```diff
--- nixos/tests/qtile.nix	2024-05-31 14:58:12.676212784 +0200
+++ nixos/tests/qtile/default.nix	2024-05-31 15:09:25.927129179 +0200
@@ -1,15 +1,46 @@
-import ./make-test-python.nix ({ lib, ...} : {
+import ../make-test-python.nix ({ lib, ...} : {
   name = "qtile";
 
   meta = {
     maintainers = with lib.maintainers; [ sigmanificient ];
   };
 
-  nodes.machine = { pkgs, lib, ... }: {
-    imports = [ ./common/x11.nix ./common/user-account.nix ];
+  nodes.machine = { pkgs, lib, ... }: let
+    qtile-config = pkgs.callPackage
+      ({ stdenvNoCC, fetchurl, tree }:
+      stdenvNoCC.mkDerivation {
+      name = "qtile-config";
+      version = "0.0.1";
+
+      src = fetchurl {
+        url = "https://raw.githubusercontent.com/qtile/qtile/master/libqtile/resources/default_config.py";
+        hash = "sha256-Y5W277CWVNSi4BdgEW/f7Px/MMjnN9W9TDqdOncVwPc=";
+      };
+
+      prePatch = ''
+        cp $src config.py
+      '';
+
+      patches = [ ./add-widget.patch ];
+
+      dontUnpack = true;
+      dontBuild = true;
+
+      installPhase = ''
+        mkdir -p $out
+        cp config.py $out/config.py
+      '';
+    }) {};
+  in {
+    imports = [ ../common/x11.nix ../common/user-account.nix ];
     test-support.displayManager.auto.user = "alice";
 
-    services.xserver.windowManager.qtile.enable = true;
+    services.xserver.windowManager.qtile = {
+      enable = true;
+      configFile = "${qtile-config}/config.py";
+      extraPackages = ps: [ ps.qtile-extras ];
+    };
+
     services.displayManager.defaultSession = lib.mkForce "none+qtile";
 
     environment.systemPackages = [ pkgs.kitty ];
@@ -30,5 +61,8 @@
         machine.wait_for_window(r"alice.*?machine")
         machine.sleep(2)
         machine.screenshot("terminal")
+
+    with subtest("ensure we can restart qtile properly"):
+        machine.succeed("pkill -f qtile -SIGUSR2")
   '';
 })

--- /dev/null	2024-05-29 08:25:18.003968520 +0200
+++ qtile/add-widget.patch	2024-05-31 14:55:47.225636465 +0200
@@ -0,0 +1,19 @@
+--- a/config.py	2024-05-31 14:49:23.852287845 +0200
++++ b/config.py	2024-05-31 14:51:00.935182266 +0200
+@@ -29,6 +29,8 @@
+ from libqtile.lazy import lazy
+ from libqtile.utils import guess_terminal
+ 
++from qtile_extras import widget
++
+ mod = "mod4"
+ terminal = guess_terminal()
+ 
+@@ -162,6 +164,7 @@
+                 # NB Systray is incompatible with Wayland, consider using StatusNotifier instead
+                 # widget.StatusNotifier(),
+                 widget.Systray(),
++                widget.AnalogueClock(),
+                 widget.Clock(format="%Y-%m-%d %a %I:%M %p"),
+                 widget.QuickExit(),
+             ],

--- all-tests.nix	2024-05-31 15:03:10.720117695 +0200
+++ all-tests.nix	2024-05-31 15:02:53.903519615 +0200
@@ -783,7 +783,7 @@
   qgis = handleTest ./qgis.nix { qgisPackage = pkgs.qgis; };
   qgis-ltr = handleTest ./qgis.nix { qgisPackage = pkgs.qgis-ltr; };
   qownnotes = handleTest ./qownnotes.nix {};
-  qtile = handleTest ./qtile.nix {};
+  qtile = handleTest ./qtile {};
   quake3 = handleTest ./quake3.nix {};
   quicktun = handleTest ./quicktun.nix {};
   quorum = handleTest ./quorum.nix {};
```
---

Using `nix-build -A qtile-unwrapped.passthru.tests.qtile`:
```
machine: must succeed: pkill -f qtile -SIGUSR2
machine # [   19.438641] xsession[973]: Current thread 0x00007fcf6d73e740 (most recent call first):
machine # [   19.440837] xsession[973]:   File "/nix/store/7hnr99nxrd2aw6lghybqdmkckq60j6l9-python3-3.11.9/lib/python3.11/selectors.py", line 468 in select
(finished: must succeed: pkill -f qtile -SIGUSR2, in 0.04 seconds)
(finished: subtest: ensure we can restart qtile properly, in 0.04 seconds)
(finished: run the VM test script, in 20.08 seconds)
machine # [   19.443664] xsession[973]:   File "/nix/store/7hnr99nxrd2aw6lghybqdmkckq60j6l9-python3-3.11.9/lib/python3.11/asyncio/base_events.py", line 1898 in _run_once
machine # [   19.446190] xsession[973]:   File "/nix/store/7hnr99nxrd2aw6lghybqdmkckq60j6l9-python3-3.11.9/lib/python3.11/asyncio/base_events.py", line 608 in run_forever
machine # [   19.449796] xsession[973]:   File "/nix/store/7hnr99nxrd2aw6lghybqdmkckq60j6l9-python3-3.11.9/lib/python3.11/asyncio/base_events.py", line 641 in run_until_complete
machine # [   19.452126] xsession[973]:   File "/nix/store/7hnr99nxrd2aw6lghybqdmkckq60j6l9-python3-3.11.9/lib/python3.11/asyncio/runners.py", line 118 in run
machine # [   19.455474] xsession[973]:   File "/nix/store/7hnr99nxrd2aw6lghybqdmkckq60j6l9-python3-3.11.9/lib/python3.11/asyncio/runners.py", line 190 in run
machine # [   19.457604] xsession[973]:   File "/nix/store/vvcadscrir6vlrpf25zlln6vydypzlxr-python3-3.11.9-env/lib/python3.11/site-packages/libqtile/core/manager.py", line 203 in loop
machine # [   19.459820] xsession[973]:   File "/nix/store/vvcadscrir6vlrpf25zlln6vydypzlxr-python3-3.11.9-env/lib/python3.11/site-packages/libqtile/scripts/start.py", line 109 in start
machine # [   19.463392] xsession[973]:   File "/nix/store/vvcadscrir6vlrpf25zlln6vydypzlxr-python3-3.11.9-env/lib/python3.11/site-packages/libqtile/scripts/main.py", line 77 in main
machine # [   19.466019] xsession[973]:   File "/nix/store/pn9nnil1g0nvkal3xrcb89cxs65yxzp0-python3.11-qtile-0.26.0/bin/.qtile-wrapped", line 9 in <module>
```

**Comment by Rutherther** [comment]
Great, thanks for the test. I am not sure what's going on though, your test outputs an error that's completely unrelated to this issue, which can be seen from the stacktrace - for missing qtile-extras there would be stacktrace to config.py where import is. The error happens both with and without this PR applied if I am testing properly, so nothing new introduced by this.

**Comment by Sigmanificient** [comment]
> Great, thanks for the test. I am not sure what's going on though, your test outputs an error that's completely unrelated to this issue, which can be seen from the stacktrace - for missing qtile-extras there would be stacktrace to config.py where import is. The error happens both with and without this PR applied if I am testing properly, so nothing new introduced by this.

Any idea how to fix this test?

**Comment by Rutherther** [comment]
This is not related to qtile-extras, the issue can be reproduced even without that config patch

**Comment by Sigmanificient** [comment]
> This is not related to qtile-extras, the issue can be reproduced even without that config patch

I think the command check at the end is wrong, but i dont know what to replace it with

**Comment by arjan-s** [comment]
I can probably take a look at the test in the weekend.
Also, like I mentioned, I'm trying to make the qtile package provide proper desktop session files (for login managers), but this probably requires changes to the wrapping, potentially making work on this PR obsolete (in the best case) or incompatible.

**Comment by Sigmanificient** [comment]
Btw when [nixos/tests/qtile: Run only on linux (#315749)](https://github.com/NixOS/nixpkgs/pull/315749) will be merged, i think i'll pr a test with `qtile-extras`. Not restart check, but just a test to ensure it works in first place

**Comment by arjan-s** [comment]
FYI: the session files PR should now also fix the underlying issue that this PR addresses

**Comment by Rutherther** [comment]
Nice, yes, that includes basically the same code (and more of course), so it should fix this, thanks. Closing this PR.
